financial-templates-lib/GasEstimator.js

```diff
@@ -0,0 +1,44 @@
+// This script gets the current recommended `fast` gas price from ethGasStation
+// to inform the Liquidator and dispute bot of a reasonable gas price to use.
+
+const fetch = require("node-fetch");
+const url = "https://ethgasstation.info/json/ethgasAPI.json";
+
+// If no updateThreshold is specified then default to updating every 60 seconds.
+class GasEstimator {
+  constructor(updateThreshold = 60) {
+    this.updateThreshold = updateThreshold;
+    this.lastUpdateTimestamp;
+    this.lastFastPriceGwei = 20; // default until updated by the `_update` function.
+  }
+
+  _update = async () => {
+    const currentTime = Math.floor(Date.now() / 1000);
+    if (currentTime > this.lastUpdateTimestamp + this.updateThreshold) {
+      return;
+    } else {
+      let response = await this.getData(url);
+      // The number returned by EthGasStation is a Gwei amount, scaled by 10.
+      this.lastFastPriceGwei = response.fast / 10;
+    }
+  };
+
+  // Returns the current fast gas price in Wei, converted from the stored Gwei value.
+  getCurrentFastPrice = () => {
+    return this.lastFastPriceGwei * 1e9;
+  };
+
+  getData = async url => {
+    try {
+      const response = await fetch(url);
+      const json = await response.json();
+      return json;
+    } catch (error) {
+      console.log(error);
+    }
+  };
+}
```

Comment:
When you call this function, you assume it will return a value, but in the catch case, it doesn't. This will cause the caller to error when it tries to lookup the `undefined.fast`. Is that desired?

Reply:
Ye, this is a good point. I updated this function to return an explicit price and deal with the error case more gracefully.

> > > > Open question: what happens if the EthGasStation API goes down, responds slowly or blocks us. We need a more sophisticated way to deal with this than hanging the bot. For now the 1 min rate limiting function will deal with this sufficiently but a more sophisticated measure should be implemented in the future.
> > > 
> > > 
> > > I see you have set a default (i.e. 20 gwei), wouldn't it revert to default? If so, why is that not a valid backup?
> > 
> > 
> > I think it should but want to get feedback on this. It is a primary failure mode as a result of the introduced dependency.
> 
> I think if this default works properly, I see no reason not to set a default (I'd make the default higher though). Is there any reason to suspect that average gas prices have increased over time?

I agree. it should be larger. I made it 40 Gwei